src/query_graph/extract_subgraphs_from_supergraph.rs

```diff
@@ -2067,3 +2067,769 @@ fn maybe_dump_subgraph_schema(subgraph: FederationSubgraph, message: &mut String
         ),
     };
 }
+
+#[cfg(test)]
+mod tests {
+    use apollo_compiler::{name, Schema};
+
+    use crate::{schema::FederationSchema, ValidFederationSubgraphs};
+
+    // JS PORT NOTE: these tests were ported from
+    // https://github.com/apollographql/federation/blob/3e2c845c74407a136b9e0066e44c1ad1467d3013/internals-js/src/__tests__/extractSubgraphsFromSupergraph.test.ts
+
+    #[test]
+    fn handles_types_having_no_fields_referenced_by_other_objects_in_a_subgraph_correctly() {
+        /*
+         * JS PORT NOTE: the original test used a Federation 1 supergraph.
+         * The following supergraph has been generated from:
+
+            federation_version: =2.6.0
+            subgraphs:
+                a:
+                    routing_url: http://a
+                    schema:
+                        sdl: |
+                            type Query {
+                                q: A
+                            }
+
+                            type A {
+                                a: B
+                            }
+
+                            type B {
+                                b: C @provides(fields: "c")
+                            }
+
+                            type C {
+                                c: String
+                            }
+                b:
+                    routing_url: http://b
+                    schema:
+                        sdl: |
+                            type C {
+                                c: String
+                            }
+                c:
+                    routing_url: http://c
+                    schema:
+                        sdl: |
+                            type D {
+                                d: String
+                            }
+
+            pbpaste | rover supergraph compose --config - | pbcopy
+
+        * The general idea being that all types are "value types" so ends with no 'join__type' in the supergraph (so
+        * `extractSubgraphsFromSupergraph` will have to initially assume that all types are in all subgraphs), but
+        * due to the `@provides`, `B.b` has a `join__field`, which makes it be only in "ServiceA". As as result,
```

Comment:
I thought `@provides` is only applicable on the entities? Was that supergraph v1 quirk?

Reply:
I honestly didn't read this comment completely — this is absolutely a supergraph v1 note. "will have to initially assume that all types are in all subgraphs" isn't true in v2 when all types are annotated with `join__type` (you can see how v1 supergraphs [leave off annotations here](https://github.com/apollographql/federation/blob/main/internals-js/src/__tests__/extractSubgraphsFromSupergraph.test.ts#L57-L75).)

Reply:
we are not planning to support v1 supergraphs so unsure whether we should be porting those tests over?

Reply:
i ported over the tests I could convert to fed2. i can change this note to make that clearer. 

i tried converting all of them but:
* one test always failed — the test was asserting that extraction was doing the "wrong" thing
* two tests won't compose in fed 2

Reply:
but after going through this process, if you decided these tests are worth porting, i would not be offended

Reply:
I removed this test, but will keep the other ones.